### DISCUSSION TITLE
When Sqlboiler generates models, it has missed composite unique keys

### DISCUSSION THREAD

**Discussion by suvaytranh** [aarondl/sqlboiler#1352] [Q&A] [answered]
Currently, when I have a table with 4 columns such as ABC(x, y, h, z), I specified constrains as the following

- Primary key: x
- Unique key: y
- Composite unique keys: h,z

After running migration and re-generating this model by SQL boiler command, I discovered that it didn't generate the composite unique keys (h,z) information at this model. It only has the information of the unique key y. I am afraid that it won't work properly if we use the Upsert function in the model to insert/update data in database. Is there anyone knows the root cause? and how can I fix it?

**Comment by stephenafamo** [comment]
SQLBoiler does not need to know about the unique keys for the upsert method to work since the query will be passed on the the DB.

Or am I missing something?

**Reply by suvaytranh** [reply]
@stephenafamo , thanks for your response. I would like to confirm more about this. 
If so, lacking the unique key of h and z at below generated code after generating models doesn't cause any problems, does it?
```
var mySQLABCUniqueColumns = []string{
	"x",
	"y",
}
```
The above code is generated after I ran sqlboiler command to generate this model.

**Reply by stephenafamo** [reply, marked as answer]
I see this is specific to MySQL upserts. I was thinking of the Postgres upsert method.  
Unfortunately, SQLBoiler does not have good support for composite keys so supporting this would be a big change.

If you want to take a look and send in a PR, that would be great. It would likely also make it easier to support composite foreign keys.

**Comment by aganguly-godaddy** [comment]
Hi @stephenafamo. I have a similar problem and realised we don't do composite unique constraints in sqlboiler. A little question: may I please know what was the reason the decision to not allow composite unique constraints? Just out of curiosity.

**Reply by stephenafamo** [reply]
It is an original limitation in how SQLBoiler was implemented.  
While it is possible to fix this, it will be quite involved and likely breaking.
